File: autorun.js
const childProcess = require("child_process");
const process = require("process");
const fs = require("fs");


var configFile = fs.readFileSync("./config.json", "utf8");
var config = JSON.parse(configFile);

const DEBUG = config.enabledebug;
function debugConsoleLog(text, text2) {
    if (DEBUG === true) {
        console.log("\033[32m[DEBUG]\033[0m", text, (text2 == undefined) ? "" : text2);
    };
};
debugConsoleLog("ENABLED");
debugConsoleLog("config.json:", config);

var timeList = new Map(config.timeList);
var playCmd = config.playCmd;
var processKillCmd = config.processKillCmd;
var timeOut = config.timeOut;
timeOut = (Number(timeOut) < 1000 || Object.is(Number(timeOut),NaN)) ? 1000 : Number(timeOut);
debugConsoleLog("timeList:", timeList);
debugConsoleLog("timeOut:", timeOut);


var targetTime;
var nith = false;
let now = new Date();
let day = now.getDay();


function play() {
    childProcess.exec(playCmd, () => {
        console.log("子进程已结束");
    });
};

function again() {
    if (processKillCmd === false) {
        run();
    } else {
        childProcess.exec(processKillCmd, (err) => {
            console.log((err) ? "\033[31m[WARN]\033[0m " + "子进程关闭失败" : "子进程关闭成功");
            run();
        });
    };
};

function getTargetTime() {
    let d, h, m, sh, sm, now = new Date();
    nith = false, d = now.getDay(), h = now.getHours(), m = now.getMinutes();
    chooseSh: for (let key of timeList) {
        if (key[0] >= h) {
            sh = key[0];
            debugConsoleLog("sh =", sh);
            let minList = timeList.get(sh);
            debugConsoleLog("minList =", minList);
            for (let key of minList) {
                if ((m < key || sh - h > 0) || nith == true) {
                    sm = key;
                    break chooseSh;
                };
            };
            debugConsoleLog("nith =", true);
            nith = true;
        };
    };
    debugConsoleLog("sm =", sm);
    if (sm == undefined) {
        console.log("\033[31m[WARN]\033[0m " + "目标时间均已过,脚本停止");
        process.exit();
    };
    debugConsoleLog("target: " + ((sh > 9) ? sh : "0" + sh) + ":" + ((sm > 9) ? sm : "0" + sm));
    return sh + ":" + sm;
};

function getleftTime(setTime) {
    let now = new Date(), h = now.getHours(), m = now.getMinutes();
    let x = setTime.split(":")[0], y = setTime.split(":")[1];
    return 60 * (Number(x) - h) + (Number(y) - m);
}

function main() {
    let sh = targetTime.split(":")[0], sm = targetTime.split(":")[1];
    let leftTime = getleftTime(targetTime);
    let now = new Date(),
        h = now.getHours(),
        m = now.getMinutes(),
        s = now.getSeconds();
    if (leftTime == 0) {
        console.log("时间到, 启动子进程");
        play();
        console.log("子进程启动,脚本暂停", timeOut, "毫秒");
        setTimeout(again, timeOut);
    } else {
        let output = "现在是";
        output += (h < 10) ? "0" + h + ":" : h + ":";
        output += (m < 10) ? "0" + m + ":" : m + ":";
        output += (s < 10) ? "0" + s : s;
        output += ",目标";
        output += ((sh > 9) ? sh : "0" + sh) + ":" + ((sm > 9) ? sm : "0" + sm) + ",剩余" + leftTime + "min";
        console.log(output);
        setTimeout(main, (leftTime > 1) ? 60000 : 1000);
    };
};

function run() {
    targetTime = getTargetTime();
    main();
};

run();

File: autorun.py
import json
import os
import time


with open("./config.json", "r", encoding="utf8") as f:
    config = json.load(f)

DEBUG = config["enabledebug"]


def debugConsoleLog(text, text2=None):
    if DEBUG == True:
        print("\033[32m[DEBUG]\033[0m", text, text2 if text2 else "")


debugConsoleLog("ENABLED")
debugConsoleLog("config.json:", config)

timeList = config["timeList"]
playCmd = config["playCmd"]
processKillCmd = config["processKillCmd"]
timeOut = config["timeOut"]/1000
timeOut = 1 if (timeOut < 1 or timeList == None) else timeOut
debugConsoleLog("timeList:", timeList)
debugConsoleLog("timeOut:", timeOut)

nith = False
now = time.localtime()
day = now[6]


def getTargetTime():
    sh = None
    sm = None
    nith = False
    h, m = time.localtime()[3:5]
    breakAll = False
    for key in timeList:
        if key[0] >= h:
            sh = key[0]
            debugConsoleLog("sh =", sh)
            minList = timeList[timeList.index(key)][1]
            debugConsoleLog("minList =", minList)
            for key in minList:
                if m < key or sh - h > 0 or nith == True:
                    sm = key
                    breakAll = True
                    break
            if breakAll == True:
                break
            debugConsoleLog("nith =", True)
            nith = True
    if sm == None:
        print("\033[31m[WARN]\033[0m 目标时间均已过, 脚本停止")
        exit()
    debugConsoleLog(
        f"target: {sh if sh > 9 else 0}{sh}: {sm if sm > 9 else 0}{sm}")
    return f"{sh}:{sm}"


def getleftTime(setTime):
    h, m = time.localtime()[3:5]
    x, y = map(int, setTime.split(":"))
    return 60 * (x - h) + (y - m)


def main():
    sh, sm = map(int, targetTime.split(":"))
    leftTime = getleftTime(targetTime)
    h, m, s = time.localtime()[3:6]
    if leftTime == 0:
        print("时间到, 启动子进程")
        os.system(playCmd)
        print("子进程已结束, 脚本暂停", timeOut, "秒")
        time.sleep(timeOut)
        run()
    else:
        output = "现在是"
        if h < 10:
            output += "0" + str(h) + ":"
        else:
            output += str(h) + ":"
        if m < 10:
            output += "0" + str(m) + ":"
        else:
            output += str(m) + ":"
        if s < 10:
            output += "0" + str(s)
        else:
            output += str(s)
        output += ", 目标"
        output += (str(sh) if sh > 9 else "0" + str(sh)) + \
            ":" + (str(sm)if sm > 9 else "0" + str(sm)) + \
            ", 剩余" + str(leftTime) + "min"
        print(output)
        time.sleep(60 if leftTime > 1 else 1)
        main()


def run():
    global targetTime
    targetTime = getTargetTime()
    main()


run()
